README.md
# pyscript-ruuvi-gateway

Python script for the pyscript addon in Home Assistant that watches a MQTT topic for messages from a [Ruuvi Gateway](https://ruuvi.com), decodes and processes them, and sends the data back to MQTT formatted for Home Assistant auto-discovery.

## Installation and Use

### Configuring Home Assistant
1. Install the [pyscript addon](https://hacs-pyscript.readthedocs.io/en/latest/)
2. Copy ruuvi-gateway.py to /config/pyscript/apps
3. Add ruuvi_decoders (the [ruuvi_decoders Python module](https://github.com/ruuvi-friends/ruuvi-decoders)) to requirements.txt in the /config/pyscript folder.
4. Add your application configuration to config/configuration.yaml (see the [configuration-sample.yaml](configuration-sample.yaml)) file for an example)
5. Reload the pyscript addon

### Configuring the Ruuvi Gateway
1. Select *Custom server* under Advanced when prompted to configure where to send Ruuv Tag data
2. Select MQTT and configure it as follows:
   * Server - the IP address (or resolvable FQDN) of your MQTT broker
   * Port - the port your MQTT broker is listening for non-SSL requests on; usually this is 1883
   * User name - the user to use when connecting to your MQTT broker
   * Password - the password to use when connecting to your MQTT broker
   * Client name - the client name used when connecting to the MQTT server (MAC address default is fine)
   * Topic prefix - the high-level topic prefix must match what's configured for MQTT_TOPIC_PREFIX in the script
3. Once finished, power cycle the gateway to begin sending any received sensor beacons to MQTT

### Measurements
The measurements included in the the sample configuration YAML file are:
- temperature (demonstrates custom unit and conversion to Fahrenheit using transform)
- humidity
- battery (demonstrates custom unit and calculation of percentage using transform)
- pressure
- tx_power

I specifically did not add examples for the acceleration measurements, as there's no device class for acceleration in Home Assistant currently.

#### Basic Metric Transformations
You can use a rudimentary arithmetic expression to modify a metric value by setting the value of the "transform" key for the measurement in the config file. This is best-effort; i.e. it works for the two simple use-cases I had (convert C to F, and calculate a rough (read: wildly inaccurate) battery level remaining percentage. I took the code directly from [Stack Overflow](https://stackoverflow.com/questions/2371436/evaluating-a-mathematical-expression-in-a-string); it's pretty much guaranteed to be fragile. Standard caveats apply, YMMV, etc.. 

### Troubleshooting
One tool that was extremely helpful for me during debugging (assuming you're using the Mosquitto MQTT broker addon for Home Assistant) is *mosquitto_sub* (installed by default with Mosquitto). 

E.g. to view messages being sent from the gateway to the MQTT top-level topic "ruuvi", Open an SSH terminal to your Home Assistant instance and enter:
```
mosquitto_sub -h <MQTT_BROKER_IP> -u <MQTT_USERNAME> -P "<MQTT_PASSWORD>" -v -t "ruuvi/#"
```

This will subscribe you to the topic and allow you to view messages as they come in to the broker. To check whether the messages are being correctly processed and submitted back to MQTT, replace "ruuvi" in the topic with "homeassist/sensor".

You can also add logging commands such as
```
"log.info(f"message with variable {varname}")
```

to the Python script itself and view them in the core log. 




CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.2] - 9/21/2021
### Changed
- Updated global decimal precision to be per-measurement
- Script now returns an integer if precision is 0
- Generalized battery min/max to range values in per-measurement configs
### Added
- support for RSSI
## [0.1.1] - 9/18/2021
### Changed
- Bug-fixes related to unicode in YAML, updated sample measurements config to include pressure and tx_power. 

## [0.1.0] - 9/18/2021
### Added
- The script now creates an "attributes" topic in MQTT for each ruuvitag

## [0.0.3] - 9/16/2021
### Changed
- All configuration has been moved to the main HA /config/configuration.yaml file (sample file added)
### Added
- Log output with MAC address for beaconing tags that aren't in RUUUVITAG (rudimentary discovery)
## [0.0.2] - 9/15/2021
### Changed
- Minor bug fixes
## [0.0.1] - 9/14/2021
### Added
- Initial commit



ruuvi-gateway.py
#----------------------------------------------------------------------------
# Author  : Aaron Finney
# Created : September 14, 2021
# Updated : September 21, 2021
# Version : 0.1.2
# 
# This script watches an MQTT topic for messages from a Ruuvi Gateway, decodes
# them, performs basic processing (if configured), and re-emits them to MQTT
# in the correct format for Home Assistant to consume as auto-discovered sensors.
#
# ---------------------------------------------------------------------------

from ruuvi_decoders import Df3Decoder, Df5Decoder
from json import dumps
import ast
import operator

def sensor_exists(mac):
    try:
        sensor = f"sensor.{mac}"
        return state.get(sensor) != "unavailable"
    except (NameError):
        return False

def eval_(node):
    operators = {ast.Add: operator.add, ast.Sub: operator.sub, ast.Mult: operator.mul,
                ast.Div: operator.truediv, ast.Pow: operator.pow, ast.BitXor: operator.xor,
                ast.USub: operator.neg}

    if isinstance(node, ast.Num): # <number>
        return node.n
    elif isinstance(node, ast.Constant): # <constant>
        return node.n
    elif isinstance(node, ast.BinOp): # <left> <operator> <right>
        return operators[type(node.op)](eval_(node.left), eval_(node.right))
    elif isinstance(node, ast.UnaryOp): # <operator> <operand> e.g., -1
        return operators[type(node.op)](eval_(node.operand))
    else:
        raise TypeError(node)


@mqtt_trigger(f"{pyscript.app_config['mqtt_topic_prefix']}/#")
def mqtt_msg(topic=None, payload=None, payload_obj=None):
    ruuvitags = pyscript.app_config['ruuvitags']
    measurement_config = pyscript.app_config['measurements']
    # get the MAC from the topic because data format 3 doesn't include it
    raw_mac = topic.split('/')[-1]
    mac = raw_mac.lower().replace(':','')
    if mac not in ruuvitags:
        # this sensor is not in the list of tags to process, ignore
        log.warning(f"MAC addres {mac} not in ruuvitags, ignoring tag")
        return
    if payload_obj.get("data") is not None:
        try:
            raw_data = payload_obj.get("data")
            clean_data = raw_data.split("FF9904")[1]

            # decode the data
            format = clean_data[0:2]
            data = {}
            if format == "03":
                decoder = Df3Decoder()
                data = decoder.decode_data(clean_data)
            else:
                decoder = Df5Decoder()
                data = decoder.decode_data(clean_data)
            # add rssi to the data structure
            data['rssi'] = payload_obj.get("rssi")

            for m_type in measurement_config:
                value = data.get(measurement_config[m_type]["source_metric"])
                if "transform" in measurement_config[m_type]:
                    value = eval_(ast.parse(f"{measurement_config[m_type]['transform']['expression']}".replace("<VALUE>",str(value)), mode='eval').body)
                    if "min_val" in measurement_config[m_type]['transform']:
                        value = min(value, measurement_config[m_type]['transform']['max_val'])
                    if "max_val" in measurement_config[m_type]['transform']:
                        value = max(value, measurement_config[m_type]['transform']['min_val'])

                if isinstance(value,float):
                    if "precision" in measurement_config[m_type]:
                        value = round(value, measurement_config[m_type]['precision'])
                        if measurement_config[m_type]['precision'] == 0:
                            value = int(value)

                if not sensor_exists(mac):
                    config_topic = f"homeassistant/sensor/{mac}/{m_type}/config"
                    config_payload = dumps(measurement_config[m_type]['config']).replace('<MAC>',mac).replace('<TAG_NAME>', ruuvitags[mac])
                    attributes_topic = f"homeassistant/sensor/{mac}/{m_type}/attributes"
                    attributes_payload = dumps(measurement_config[m_type]['attributes']).replace('<MAC>',mac)
                    mqtt.publish(topic=config_topic, payload=config_payload, retain='True')
                    mqtt.publish(topic=attributes_topic, payload=attributes_payload, retain='True')

                state_topic = f"homeassistant/sensor/{mac}/{m_type}/state"
                mqtt.publish(topic=state_topic, payload=value, retain='True')

        except (AttributeError, ValueError, TypeError) as err:
            log.warning(f'Error --> {payload}')
            log.warning(f'Error code -> {err}')
